path_login.go

```diff
@@ -0,0 +1,301 @@
+// Copyright © 2019, Oracle and/or its affiliates.
+package ociauth
+
+import (
+	"context"
+	"fmt"
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/oracle/oci-go-sdk/common"
+	"github.com/pkg/errors"
+	"net/http"
+	"strings"
+	"time"
+	"unicode"
+)
+
+// These constants store the required http path & method information for validating the signed request
+const (
+	PathVersionBase = "/v1"
+	PathBaseFormat  = "/auth/oci/login/%s"
+	PathLoginMethod = "get"
+)
+
+// Signing Header constants
+const (
+	// HdrRequestTarget represents the special header name used to refer to the HTTP verb and URI in the signature.
+	HdrRequestTarget = `(request-target)`
+	HdrHost          = `host`
+)
+
+func pathLogin(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "login/" + framework.GenericNameRegex("role"),
+		Fields: map[string]*framework.FieldSchema{
+			"requestHeaders": {
+				Type:        framework.TypeMap,
+				Description: `The signed headers of the client`,
+			},
+			"role": {
+				Type:        framework.TypeString,
+				Description: "Name of the role.",
+			},
+		},
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLoginUpdate,
+		},
+
+		HelpSynopsis:    pathLoginSyn,
+		HelpDescription: pathLoginDesc,
+	}
+}
+
+func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	//Validate the role
+	role, ok := data.GetOk("role")
+	if !ok {
+		return unauthorizedLogicalResponse(req, b.Logger(), fmt.Errorf("role is not specified"))
```

Comment:
This helper function is probably not needed, and is behaving a little different than normal Vault auth. The error message is only being debug logged and not returned to the user trying to log in. If the error is due to basic user error, that is typically returned in order to help the user troubleshoot. For example I'd expect to see here:
```
return logical.ErrorResponse("role is not specified"), nil
```